Add wait subsystem for Unix

kouvel opened the pull request:
- See comment at the top of WaitSubsystem.Unix.cs for details
- Refactored WaitHandle and derived classes, and RuntimeThread, to separate platform-specific functionality
- These changes have only been tested to the extent of public APIs currently available. They have not yet been tested with multiple threads, and several features such as sleep, interrupt, signal-and-wait, mutex abandoning, etc., have not yet been tested. More tests will be coming after Thread is made available in CoreRT.

dnfclas commented:
Hi __@kouvel__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

kouvel commented:
@dotnet-bot test this please

jkotas commented:
@AntonLapounov PTLA

jkotas commented on src/System.Private.CoreLib/src/System/Threading/RuntimeThread.cs:
```diff
@@ -2,33 +2,123 @@
+using System.Runtime;
+using Microsoft.Win32.SafeHandles;
 
 namespace System.Threading
```
Could you please move the RuntimeThread into same namespace as it is on CoreCLR, so that we do not have to ifdefs in corefx? (There are some already that will need to be removed.)

kouvel replied:
Will do, I'll also move the files to System.Private.CoreLib\src\Internal\Runtime\Augments

AntonLapounov replied:
I don't think we can move the class without breaking ProjectN.  Can we keep the old S.T.RuntimeThread class and forward from it to the new I.R.A.RuntimeThread?

kouvel replied:
Sure, that should work

jkotas replied:
@AntonLapounov What does depend on it in ProjectN today?

AntonLapounov replied:
My threading work for ProjectN.  The official build won't be affected.

jkotas replied:
So you can add the temporary wrapper locally to keep you going; and we can have the actual implementation in the right place so that it starts propagating through the system.

kouvel replied:
Since an instance of RuntimeThread needs to be I.R.A.RuntimeThread, which Thread derives from, I.R.A.RuntimeThread would need to derive from the temporary S.T.RuntimeThread to forward instance members without an extra indirect and the main functionality would need to temporarily remain in S.T.RuntimeThread. Or is there another way? Wouldn't it be easier to just target I.R.A.RuntimeThread directly from your changes?

AntonLapounov replied:
I am simply asking to not remove the old file so I can use it as a wrapper.  If the file is deleted, I'll have to recreate it until we get an updated package from CoreFX.  The actual implementation should go into the right place, I.R.A.RuntimeThread.

kouvel replied:
Ok sounds good

jkotas replied:
>  I.R.A.RuntimeThread, which Thread derives from

Nothing should be deriving from RuntimeThread in CoreRT. Correct?

jkotas replied:
We should mark RuntimeThread as sealed in CoreRT to make it clear.

kouvel replied:
Nothing should derive from it in CoreRT, I didn't know that it can be marked as sealed and have Thread from CoreFX still derive from it. Will fix.

jkotas replied:
CoreFX Thread wraps it, it does not derive from it.

https://github.com/dotnet/corefx/blob/277b781b6326e06716173132f68317ddc475f67d/src/System.Threading.Thread/src/System/Threading/Thread.cs#L32

kouvel replied:
My bad, got some things confused

kouvel replied:
Made the change. Chatted with @AntonLapounov offline, removed RuntimeThread from System.Threading to avoid name conflicts that would require renaming all uses inside System.Threading temporarily.

kouvel commented:
CC @sergiy-k as well

Due to the refactoring, for reviewing it may be easier to pull down the changes and review using your favorite editor / diff tool.

kouvel commented:
@dotnet-bot test Windows_NT Debug and CoreCLR tests

Failure in BasicThreading seems to be unrelated, I'm not seeing it locally

jkotas commented on src/Native/System.Private.CoreLib.Native/pal_threading.cpp:
```diff
@@ -0,0 +1,134 @@
+    error = pthread_condattr_init(&conditionAttributes);
+    if (error != 0)
+    {
+        throw OutOfMemoryException();
```
We should not be using C++ exceptions in the CoreLib PAL. It creates dependency on C++ runtime that is a pain. None of the other System.*.Native or CoreRT runtime are using C++ exceptions.

kouvel replied:
Fixed

sergiy-k commented:
@kouvel, the failure in the BasicThreading test happens when the test is built and run in the multi-module mode. I can easily reproduce it with "tests\runtest.cmd Debug /multimodule"
It crashes in GC while scanning handle tables:

```
BasicThreading!WKS::gc_heap::mark_object_simple1+0x421
BasicThreading!WKS::gc_heap::mark_object_simple+0x2da
BasicThreading!WKS::GCHeap::Promote+0x1fd
BasicThreading!PromoteObject+0x76
BasicThreading!ScanConsecutiveHandlesWithoutUserData+0xb3
BasicThreading!BlockScanBlocksWithoutUserData+0x71
BasicThreading!SegmentScanByTypeChain+0xeb
BasicThreading!TableScanHandles+0x128
BasicThreading!HndScanHandlesForGC+0x2a9
BasicThreading!Ref_TraceNormalRoots+0x1a9
BasicThreading!GCScan::GcScanHandles+0xa2
BasicThreading!WKS::gc_heap::mark_phase+0x748
BasicThreading!WKS::gc_heap::gc1+0x16e
BasicThreading!WKS::gc_heap::garbage_collect+0x897
BasicThreading!WKS::GCHeap::GarbageCollectGeneration+0x428
BasicThreading!WKS::GCHeap::GarbageCollectTry+0xf5
BasicThreading!WKS::GCHeap::GarbageCollect+0x345
BasicThreading!RhpCollect+0x94
BasicThreading!RhCollect+0x3f
BasicThreading!System_Private_CoreLib_System_GC__Collect_0+0x1a
BasicThreading!BasicThreading_FinalizeTest__Run+0x43
BasicThreading!BasicThreading_Program__Main+0x36
```
The failure does not repro if I change the test to execute FinalizeTest before the test case that creates Tasks (i.e. before ThreadStaticsTestWithTasks). So there is a chance that it might be related your changes in *WaitHandle classes but I'm not 100% sure. It is also possible that you change just exposed an existing issues.

jkotas commented on src/System.Private.CoreLib/src/System/Threading/LowLevelThread.cs:
```diff
@@ -64,30 +62,77 @@ internal static unsafe int WaitForMultipleObjects(IntPtr* pHandles, int numHandl
+                    /// SignalObjectAndWait function does not return an error.
+                    throw new SynchronizationLockException();
+                    // TODO: netstandard2.0 - After switching to ns2.0 contracts, use the below instead for compatibility
+                    //throw new ApplicationException(SR.Arg_SynchronizationLockException);
```
ApplicationException was added already.

kouvel replied:
Looks like it's there in S.P.CoreLib but I can't test for it yet because it's not yet in the API surface used by the tests

jkotas replied:
Updating the tests to latest contracts would probably require fist fight with nuget ... not worth it.

You should be able to test for the ApplicationException HResult ... better than nothing.

kouvel replied:
That works, will fix

jkotas commented on src/System.Private.CoreLib/src/System/Threading/WaitSubsystem.HandleManager.Unix.cs:
```diff
@@ -0,0 +1,125 @@
+            {
+                s_lock.VerifyIsLocked();
+
+                WaitableObject waitableObject = FromHandle_NoThrow(handle.ToInt32() - 1);
```
Note that ToInt32 throws ... if you want real no throw version,  just use cast.

kouvel replied:
I should do a check and throw the typical invalid handle exception if it's out of range, and use a cast afterwards

jkotas commented on src/System.Private.CoreLib/src/System/InvalidOperationException.cs:
```diff
@@ -38,5 +38,12 @@ public InvalidOperationException(String message, Exception innerException)
 
         protected InvalidOperationException(SerializationInfo info, StreamingContext context) : base(info, context) { }
+
+        internal static InvalidOperationException NewInvalidHandle()
```
This should be in the handle manager where it is needed, not here.

kouvel replied:
It's used in several places, but I noticed now that there is a SetErrorCode to set the HResult, I can move it, maybe SafeHandle is a good place for it

jkotas replied:
Ok. Feel free to ignore this comment then.

jkotas commented on src/System.Private.CoreLib/src/System/Threading/WaitSubsystem.HandleManager.Unix.cs:
```diff
@@ -0,0 +1,125 @@
+
+            private static WaitableObject[] s_objects = new WaitableObject[InitialCapacity];
+            private static int s_count;
+            private static LowLevelQueue<int> s_freeIndexes;
```
Having free list that is parallel with `WaitableObject[]` may be better.  A lot less code than Queue, not out of memory on attempt to free the code, easier to detect bad handles.

- When the handle index is free, it would be index of next free element.
- When the handle index is occupied, it would be "cookie" to detect invalid handles. Something like:

```
struct Handle
{
    public WaitableObject Value;
    public int Link;
}

static Handle[] s_handles = ...


public static WaitableObject FromHandle(IntPtr handle)
{
     int index = ((int)handle - 1) & (s_handles.Length - 1);
     ref Handle handle = s_handles[index];
     
     WaitableObject value = handle.Value

     if (value == null || handle.Link != (int)handle)
          throw InvalidOperationException.NewInvalidHandle(); 

     return value;
}
```

kouvel replied:
When a handle is deleted, how would I know what the next free handle index is, to fill that slot? I can't think of a way without walking the array or other more complicated solutions.

The OOM on delete issue can be fixed by adjusting the queue capacity when the array capacity is adjusted. I just didn't think it's worth it, since after the live handle count stabilizes, the queue's backing array could be fairly small if there are no large bursts of temporary handles. But it could go either way.

jkotas replied:
Classic free list:

```
int s_freeList;
...

   s_handles[deletedIndex].Link = s_freeList;
   s_freeList = deletedIndex;
```

kouvel replied:
With that, it seems like the most recently deleted indexes would be reused first. I wanted to keep deleted indexes free for some time such that we can detect uses of deleted handles before they are reused.

kouvel replied:
Maybe something like this could work:

```
int s_deletedHandleHeadIndex;
int s_deletedHandleTailIndex;
...

    assert(s_handles[s_deletedHandleTailIndex] == -1); // end of queue
    s_handles[s_deletedHandleTailIndex] = deletedIndex;
    s_handles[deletedIndex] = -1;
    s_deletedHandleTailIndex = deletedIndex;
```

Then it can maintain queue-style info with your data structure. I'll give it a try.

jkotas replied:
The upper bits of each new handle can be filled with random bits. The `rec.Link != handle` check in FromHandle would detect use of deleted handles because of the upper random bits of the deleted handle won't match upper random bits of the current handle.

Allocation of new handle:
```
int index = s_freeList;
if (index == -1) ...
s_freeList = s_handles[index].Link;

int mask = (s_handles.Length - 1) & 0x3FFFFFFF;
IntPtr handle = s_handles[index] = ((random.Next() & ~mask) | index) + 1;
```

jkotas replied:
> Then it can maintain queue-style info 

Yes, that would work too.

kouvel replied:
Taking a step back, it looks like this would mainly avoid a Queue, and I agree a checksum would add additional verification on top of a queue. A checksum on its own without a queue may have collision issues depending on how many high bits are used for the checksum.

I'm trying to rationalize the need for this change. Do you think it's worth making this change?

jkotas replied:
All this is just a best effort to help catch bugs. Neither queue or checksum guarantee anything.

I do not think that couple hundred lines of the copied corefx queue implementation are the price worth paying for it. My primary motivation for the comment was whether similar effect can be achieved with much less code.

kouvel replied:
Well, less code, yes but copied code in a sense is less code, as I hope we would share the code eventually such that it would not be a separate copy eventually.

kouvel replied:
But I'm still trying to understand the core of your motivation. Why is less code a good thing if it's more complicated?

jkotas replied:
I do not think we would want to move the Queue from corefx to corelib.

kouvel replied:
I understand that, but if that leads to design changes, isn't there a problem? What if I need a queue in CoreLib for perfectly acceptable purposes?

kouvel replied:
There is already a LowLevelList, I assume for similar reasons

jkotas replied:
> LowLevelList

And LowLevelListWithIList, LowLevelDictionary, LowLevelDictionaryWithIEnumerable, ConcurrentUnifier, ConcurrentUnifierW, ConcurrentUnifierWKeyed, LowLevelListDictionary, ... there many broken collection clones. That's why I am not thrilled about adding another one for what seems pretty low value. I would not be opposed to adding one when it is justified.

I believe that something like handle manager is best written directly against metal and kept as simple possible.

From what I can tell, CoreCLR PAL handle manager has a simple free list: https://github.com/dotnet/coreclr/blob/32f0f9721afb584b4a14d69135bea7ddc129f755/src/pal/src/handlemgr/handlemgr.cpp#L145. I do not remember anybody ever wishing for having a more elaborate one, and the risk for screw-ups in CoreCLR is higher because of it is a ton of C++. I am wondering why we cannot have just that here as well.

kouvel replied:
This raises a few questions:

> That's why I am not thrilled about adding another one for what seems pretty low value.

Why are basic data structures in CoreFX then, and not in CoreLib? Perhaps the reason the copies exist is simply because the originals are in a place where they can't be used considering the current separation of code between CoreLib and CoreFX. Maybe some uses of these data structures are not good uses of those data structures. I'm fine with that. But when there are good uses of these basic data structures, the only solution based on the current architecture is to copy the code to CoreLib and that does not seem like a sufficient solution. Another solution is to desperately find an alternative that is more complicated and effectively accomplishes the same thing, I'm certainly not convinced by that argument.

> I believe that something like handle manager is best written directly against metal and kept as simple possible.

I'm not sure that this motivation is simply due to the lack of the appropriate data structures being available already.

> From what I can tell, CoreCLR PAL handle manager has a simple free list

CoreCLR PAL handle manager makes no attempt to prevent recently deleted handles from being reused, and in my opinion, is by no means simple, at the very least, not simpler than the handle manager in this change, hence my resistance. Why not make it simpler, if nobody has cared before? And if there are reasons to care, I'd be happy to discuss.

jkotas replied:
What to have in a CoreLib and what to not have it is a fine balance. We did not always paid the right amount pf attention it.

Anyway, now that I slept on it I think the handle manager should be even more simpler: It should just allocate GC handles. Yes, you won't be able to detect some uses of invalid handles instead of crashing but I think it is absolutely fine.

It is very rare for managed code to operate on wait handles as IntPtrs. If there is such code and gets it wrong, it is fine for it to have different failure pattern. No need to be bug-for-bug compatible.

If we find that GCHandle lifetime is a common problem, we should build a debugging aid for it separately. It will cover this as well as all other GCHandle use cases.

kouvel replied:
It looks like that would work well, I'll go ahead with using GCHandle

kouvel commented:
Thanks @sergiy-k, I'm able to repro it now as well. It seems very unlikely that this change caused this issue. It looks like it's failing when trying to scan an invalid object in the handle table. There's nothing in this change that would cause that issue. It's more likely that this change exposed an existing issue.

Filed #2695, will reorder tests to unblock this change

jkotas commented on src/Native/System.Private.CoreLib.Native/pal_threading.cpp:
```diff
@@ -0,0 +1,166 @@
+    monitor->Wait();
+}
+
+extern "C" int CoreLibNative_LowLevelMonitor_TimedWait(LowLevelMonitor *monitor, int32_t timeoutMilliseconds)
```
int -> int32_t

jkotas commented on src/System.Private.CoreLib/src/Internal/Runtime/Augments/RuntimeThread.Windows.cs:
```diff
@@ -0,0 +1,46 @@
+
+        internal static void UninterruptibleSleep0()
+        {
+            global::Interop.mincore.Sleep(0);
```
Are these global:: prefixes actually needed?

kouvel replied:
Yea there's another Internal.Interop class

jkotas commented on src/System.Private.CoreLib/src/System/Threading/LowLevelThread.cs:
```diff
@@ -18,22 +19,16 @@ namespace System.Threading
 {
     internal static class LowLevelThread
```
Is there a point of having this class? RuntimeThread.Windows.cs maybe a better home for whatever is left here.

kouvel replied:
I didn't follow the relevance to RuntimeThread.Windows.cs, could you please elaborate?

The alternative to this class is to use LowLevelMonitor (or actually a C# LowLevelMutex without a condition) for lock purposes. LowLevelLock is significantly faster, and I don't see any reason why we should p/invoke for a lock.

jkotas replied:
This file is just static methods. RuntimeThread.Windows.cs forwards calls to here, and this calls backs to RuntimeThread. I do not see a clear split in responsibility between the two.

I did not mean to actually change anything functionality wise.

kouvel replied:
My bad, I read LowLevelLock instead of LowLevelThread (it's been a long week). I don't see any specific reason why this class needs to be there, I just left it there because it was there before. The functionality could be folded into other components. Wait-related stuff could go into WaitHandle.Windows.cs, and ReentrantWait-related stuff could be moved to RuntimeThread.

jkotas commented:
@stephentoub Could you please take a look at this as well if you get chance?

stephentoub commented on src/Common/src/Interop/Windows/mincore/Interop.Threading.cs:
```diff
@@ -0,0 +1,56 @@
+{
+    internal static partial class mincore
+    {
+        [DllImport("api-ms-win-core-synch-l1-1-0.dll", EntryPoint = "CreateEventExW", CharSet = CharSet.Unicode)]
```
We moved away from API sets in coreclr/corefx... it's correct to still use them here?

jkotas replied:
We have not done pass over CoreRT to apply the same cleanup yet ... https://github.com/dotnet/corert/issues/2241

Since you are touching this file anyway, it would be nice to give it the right shape.

AntonLapounov replied:
@jkotas I thought the convention was one PInvoke per file.  Is this file different?

jkotas replied:
This what the guidelines say about this https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/interop-guidelines.md

If multiple related DllImports will all be needed by every consumer, they may be declared in the same file, named for the functionality grouping, e.g. Interop.IOErrors.cs.

stephentoub commented on src/Common/src/System/Collections/Generic/LowLevelQueue.cs:
```diff
@@ -0,0 +1,288 @@
+namespace System.Collections.Generic
+{
+    // A simple Queue of generic objects.  Internally it is implemented as a 
+    // circular buffer, so Enqueue can be O(n).  Dequeue is O(1).
```
> so Enqueue can be O(n)

Only if it needs to grow, though, right?  Otherwise it's O(1).

kouvel replied:
Yea, just copy/pasted from CoreFX. I'm going to remove this file as it's no longer necessary.

stephentoub commented on src/Native/System.Private.CoreLib.Native/pal_threading.cpp:
```diff
@@ -0,0 +1,166 @@
+    }
+
+    error = pthread_condattr_setclock(&conditionAttributes, CLOCK_MONOTONIC);
+    assert(error == 0);
```
This is ok as debug-only because the only possible error conditions would be due to invalid inputs, and we know we're providing valid inputs?

kouvel replied:
Yea. Currently in configure.cmake only checking for the existence of CLOCK_MONOTONIC and pthread_condattr_setclock, I think that should be sufficient because CLOCK_MONOTONIC is the typical argument passed to that function to change the clock from the default.

jkotas commented on src/Common/src/Interop/Windows/mincore/Interop.Threading.cs:
```diff
@@ -0,0 +1,56 @@
+
+        internal static IntPtr OpenEvent(uint dwDesiredAccess, bool bInheritHandle, string lpName)
+        {
+            return OpenEvent(dwDesiredAccess, bInheritHandle ? 1 : 0, lpName);
```
Is there a good reason to do the manual marshalling here, but not for the other similar methods?

kouvel replied:
This code was just copy/pasted from the original file. It's probably not needed, I'll look over the calls and will clean them up.

stephentoub commented on src/System.Private.CoreLib/src/Internal/Runtime/Augments/RuntimeThread.Unix.cs:
```diff
@@ -0,0 +1,47 @@
+        public ApartmentState GetApartmentState()
+        {
+            Debug.Fail("Should not reach here because CoreFX's Thread class does not forward to this function on Unix");
+            return ApartmentState.Unknown;
```
Should these functions just throw then if we don't expect them to ever be called? e.g. ```throw null``` like we have in the refs?

kouvel replied:
Yep makes sense, FailFast sounds good

stephentoub commented on src/System.Private.CoreLib/src/Internal/Runtime/Augments/RuntimeThread.Unix.cs:
```diff
@@ -0,0 +1,47 @@
+            throw new PlatformNotSupportedException();
+        }
+
+        internal static bool ReentrantWaitsEnabled => false;
```
You could potentially make this a const or a readonly static field in this implementation, even if it's a property in the Windows one.

stephentoub commented on src/System.Private.CoreLib/src/Internal/Runtime/Augments/RuntimeThread.cs:
```diff
@@ -0,0 +1,127 @@
+            _waitInfo = new WaitSubsystem.ThreadWaitInfo(this);
+#else
+            _waitedHandles = new WaitHandleArray<IntPtr>(elementInitializer: null);
+#endif
```
Since we already have this as a partial class with Windows and Unix files, could we just have a FinishInitialize method or some such thing in those partial types which does this, rather than needing to ifdef on PLATFORM_UNIX as well?

kouvel replied:
I was thinking about that but one of the fields is readonly and has to be initialized in the constructor. I can move the other one.

stephentoub commented on src/System.Private.CoreLib/src/Internal/Runtime/Augments/RuntimeThread.cs:
```diff
@@ -0,0 +1,127 @@
+        {
+            get
+            {
+                Interlocked.MemoryBarrier(); // read the latest value
```
A Volatile.Read would be insufficient?

kouvel replied:
Volatile doesn't issue a barrier on x86/x64 (not sure about the Volatile class, but I assume it's the same) and hence would not guarantee that it will ever read the latest value, and even when it does issue a barrier it's after the read. To read the latest value a memory barrier before the read is exactly what I need. Depending on need, this may be changed to use interlocked operations later, currently only the same thread changes the value so there's no need for interlocked operations yet.

jkotas replied:
All you should need here is to ensure that the codegen will not optimize fetch of _threadState out, so that patterns like active waiting on it works fine. Volatile.Read is sufficient to do that. I do not see why you need a MemoryBarrier.

kouvel replied:
That's fine by me, the caller can deal with caching issues. I'll change it to volatile.

stephentoub commented on src/System.Private.CoreLib/src/System/Threading/EventWaitHandle.Windows.cs:
```diff
@@ -0,0 +1,120 @@
+
+        private static void VerifyNameForCreate(string name)
+        {
+            if (null != name && ((int)Interop.Constants.MaxPath) < name.Length)
```
At some point I seem to remember there being discussion about just letting the OS tell us if the name is too long, rather than trying to do our own upfront verification.  Am I remembering correctly?  Where did we land on that?

kouvel replied:
I don't recall, but sounds like a good idea to me. Do we want to make that change in CoreCLR as well?

stephentoub replied:
I was thinking of https://github.com/dotnet/corefx/issues/3740.  Looks like we agreed (at least for semaphore) it was fine to do, but haven't done it yet.

kouvel replied:
I'll do this in a separate PR, as part of that issue

stephentoub commented on src/System.Private.CoreLib/src/System/Threading/FirstLevelSpinWaiter.cs:
```diff
@@ -0,0 +1,99 @@
+            }
+            finally
+            {
+                Interlocked.Decrement(ref _spinningThreadCount);
```
Is this spinning logic ported from somewhere or is it new?

kouvel replied:
It's new, I wanted to start with a light amount of spinning and go from there. The intention was to use this in any place where some spinning is done before doing a proper wait, especially in potentially high-contention cases like the process-wide lock used in the wait subsystem.

stephentoub commented on src/System.Private.CoreLib/src/System/Threading/LowLevelLock.Unix.cs:
```diff
@@ -0,0 +1,226 @@
+                    return
+                        (state2 & LockedMask) == 0 &&
+                        Interlocked.CompareExchange(ref _state, state2 + LockedMask, state2) == state2;
+                }))
```
This is going to allocate a delegate on each WaitAndAcquire call.

kouvel replied:
Oops, will fix

stephentoub commented on src/System.Private.CoreLib/src/System/Threading/Mutex.Windows.cs:
```diff
@@ -0,0 +1,143 @@
+            }
+        }
+    }
+}
```
I assume all of this code was basically just copied from coreclr?

kouvel replied:
I believe so, I did a diff a while back and there were some small differences here and there, but nothing significant. I'll take another look and make sure.

stephentoub commented on src/System.Private.CoreLib/src/System/Threading/WaitHandle.Windows.cs:
```diff
@@ -0,0 +1,183 @@
+            IntPtr[] handles = currentThread.GetWaitedHandleArray(count);
+            for (int i = 0; i < count; i++)
+            {
+                handles[i] = safeWaitHandles[i].DangerousGetHandle();
```
Did we addref somewhere already?

kouvel replied:
The addrefs are done in WaitHandle.cs just before calling the platform-specific code

kouvel replied:
In this particular case, in ObtainSafeWaitHandles. There were some issues with the way it was doing it before, so I fixed and folded it into that function

AntonLapounov commented on src/Native/System.Private.CoreLib.Native/pal_threading.h:
```diff
@@ -0,0 +1,155 @@
+    void Acquire()
+    {
+        int error = pthread_mutex_lock(&m_mutex);
+        assert(error == 0);
```
Shouldn't we be more aggressive in case an error is returned by the OS (e.g., do fail-fast)?

kouvel replied:
For unexpected cases where security is not an issue, I think an assert suffices, that is typically how it was done in other projects I have worked on. But I'm not sure about this project's guidelines on this topic.

jkotas replied:
FWIW: We are ignoring pthread_mutex_lock result in other places in CoreRT and CoreCLR, so it should be ok to ignore it here as well ... we are consistent at least.

AntonLapounov replied:
This approach seems dangerous to me as investigating race conditions is usually costly.  However, I am OK for now; we may review this later.

AntonLapounov commented on src/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeWaitHandle.cs:
```diff
@@ -36,7 +36,11 @@ public SafeWaitHandle(IntPtr existingHandle, bool ownsHandle) : base(ownsHandle)
 
         protected override bool ReleaseHandle()
         {
+#if PLATFORM_UNIX
```
I thought the convention was to put Unix-specific code to SafeWaitHandle.Unix.cs.  Or do we allow single-line platform-specific #if's like the one here?

kouvel replied:
I just made a judgement call as it felt excessive to split files just for this

AntonLapounov commented on src/System.Private.CoreLib/src/System/Threading/SpinWait.cs:
```diff
@@ -130,11 +131,11 @@ public void SpinOnce()
+                    RuntimeThread.Sleep(0);
                 }
                 else
                 {
```
Should we use RuntimeThread.Yield() in the next line for the sake of consistency?

kouvel commented:
Made changes from above except where noted, and rebased. Also since the handle manager does not need to take the lock anymore, simplified WaitSubsystem.Unix.cs.

kouvel commented:
Perf is about 5% slower than before, but still significantly faster than Windows

jkotas commented on src/System.Private.CoreLib/src/System/Threading/WaitSubsystem.HandleManager.Unix.cs:
```diff
@@ -0,0 +1,58 @@
+
+                IntPtr handle = GCHandle.ToIntPtr(GCHandle.Alloc(waitableObject));
+                Debug.Assert(handle != new IntPtr(-1)); // SafeWaitHandle treats -1 and 0 as invalid
+                if (handle == IntPtr.Zero)
```
This is not necessary - GCHandle.Alloc throws OOM on failure

kouvel replied:
I can't seem to find it:

```c#
(GCHandle constructor)
...
            _handle = RuntimeImports.RhHandleAlloc(value, type);

            // Record if the handle is pinned.
            if (type == GCHandleType.Pinned)
                SetIsPinned();
        }
```

```c++
(HndCreateHandle)
...
    // did the allocation succeed?
    if (!handle)
    {
#ifdef FEATURE_REDHAWK
        return NULL;
#else
        ThrowOutOfMemory();
#endif
    }
```

jkotas replied:
```
        internal static IntPtr RhHandleAlloc(Object value, GCHandleType type)
        {
            IntPtr h = RhpHandleAlloc(value, type);
            if (h == IntPtr.Zero)
                throw new OutOfMemoryException();
            return h;
        }
```

jkotas replied:
Would it look better to just call RhHandleAlloc directly to not have these conversions between GCHandle and IntPtr?

kouvel replied:
Darn missed that step. Yea it would be cleaner to call the imports directly.

jkotas commented on src/System.Private.CoreLib/src/Internal/Runtime/Augments/RuntimeThread.Windows.cs:
```diff
@@ -0,0 +1,141 @@
+
+            global::Interop._APTTYPE aptType;
+            global::Interop._APTTYPEQUALIFIER aptTypeQualifier;
+            int result = global::Interop.mincore.CoGetApartmentType(out aptType, out aptTypeQualifier);
```
I am wondering whether you can do using Interop = global::Interop; at the top of the file to avoid having the prefixes everywhere.

jkotas commented on src/System.Private.CoreLib/src/System/Threading/WaitHandle.Windows.cs:
```diff
@@ -0,0 +1,298 @@
+                    /// that is locked by a different thread. Note that if the mutex is already unlocked, the Windows
+                    /// SignalObjectAndWait function does not return an error.
+                    throw new SynchronizationLockException();
+                    // TODO: netstandard2.0 - After switching to ns2.0 contracts, use the below instead for compatibility
```
Switch now here and put the workaround in the test because that's where the actual problem is?

jkotas commented on src/System.Private.CoreLib/src/System/Threading/LowLevelMonitor.Unix.cs:
```diff
@@ -0,0 +1,139 @@
+            GC.SuppressFinalize(this);
+        }
+
+        public void VerifyIsLocked()
```
Mark these with `[Conditional("DEBUG")]` so that they are not even compiled into the release builds? The empty method will get likely inlined and disappear; but it is even better if it is not there at all.

kouvel replied:
I tried that first, but it looks like it still parses the code in Release builds and tries to find variables which are declared under #if DEBUG. Had the same issue with Debug.Assert statements that have expressions using variables declared under #if DEBUG. I couldn't figure out a way to make it work so I just switched to #if DEBUG where it was problematic.

jkotas commented:
> Perf is about 5% slower than before,

What are you measuring?

kouvel commented:
I just did a quick test with something like:
```c#
manualResetEvent.Set();
manualResetEvent.WaitOne(0);
manualResetEvent.Reset();
```
In a loop just to get a sense of the overhead.

kouvel commented:
Made changes from above

jkotas commented:
What is the baseline?

kouvel commented:
Before I was getting about 6300 iterations of the loop per millisecond, after about 6000, and on Windows about 2100.

jkotas commented:
Just curious - where would you expect that this difference came from?

kouvel commented:
I'm guessing it's the p/invoke in the handle manager, but let me confirm

jkotas commented:
To get the handle value? It should be "FCall" that just dereferences the pointer - it should be cheaper than was there before.

kouvel commented:
I see, in that case that can't be the issue. Are calls into CoreLibNative also FCalls without other p/invoke overhead? I was seeing a significant difference between using LowLevelLock and LowLevelMonitor for the process-wide lock and I figured pthread mutex would have a pretty good implementation.

Anyway I tried reverting the handle manager and cleanup in WaitSubsystem and measured again with better error margins instead of eyeballing it, the difference was actually about 2%. The latest commit is producing the best numbers so far. Maybe better inlining, not sure, but it's about 4% better than before that commit.

kouvel commented:
@dotnet-bot test Ubuntu Release

AntonLapounov commented on src/System.Private.CoreLib/src/System.Private.CoreLib.csproj:
```diff
@@ -119,8 +119,15 @@
   </ItemGroup>
+  <ItemGroup Condition="'$(TargetsWindows)'=='true'">
+    <Compile Include="Internal\Runtime\Augments\RuntimeThread.Windows.cs" />
+  </ItemGroup>
```
Why don't we put it into the existing <ItemGroup Condition="'$(TargetsWindows)'=='true'">?

AntonLapounov commented on src/System.Private.CoreLib/src/System.Private.CoreLib.csproj:
```diff
@@ -119,8 +119,15 @@
+  </ItemGroup>
+  <ItemGroup Condition="'$(TargetsUnix)'=='true'">
+    <Compile Include="Internal\Runtime\Augments\RuntimeThread.Unix.cs" />
+  </ItemGroup>
```
Same as above: we may put it into the existing `<ItemGroup Condition="'$(TargetsUnix)'=='true'">`.
